### PR TITLE
Check matching static extents in View constructor

### PR DESCRIPTION
**Pull request by masterleinad**
Follow-up to https://github.com/kokkos/kokkos/pull/5188 addressing https://github.com/kokkos/kokkos/pull/5188#issuecomment-1179206669.

Fixing typo/bug in `static void run_test_contruction_from_layout_2()`, see https://github.com/kokkos/kokkos/pull/5190#discussion_r1000130527.

### CODE REVIEW

**Review by nliber — APPROVED** [review]

**Review by PhilMiller — APPROVED** [review]

**Comment by dalg24** [comment]
Retest this please

File: core/src/impl/Kokkos_ViewArray.hpp [hunk 1]
```diff
@@ -158,6 +158,12 @@ class ViewMapping<Traits, Kokkos::Array<>> {
     return m_impl_offset.m_dim.extent(r);
   }
 
+  static KOKKOS_INLINE_FUNCTION constexpr size_t static_extent(
```
Comment (dalg24) [inline core/src/impl/Kokkos_ViewArray.hpp:161]: Why/where is this needed?

**Review by dalg24 — CHANGES_REQUESTED** [review]
Makes `runtime_check_rank` less readable

File: core/src/Kokkos_View.hpp [hunk 2]
```diff
@@ -92,19 +92,38 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
+      for (size_t i = dyn_rank; i < rank; ++i)
+        if (new_extents[i] != View::static_extent(i)) {
+          ranks_matching = false;
+          KOKKOS_IF_ON_HOST(const std::string message =
+                                "Constructor for Kokkos View '" + label +
+                                "' has mismatched static extents at position " +
+                                std::to_string(i) + ". Given extent is " +
+                                std::to_string(new_extents[i]) +
+                                " but should be " +
+                                std::to_string(View::static_extent(i)) + ".\n";
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_View.hpp:119]: Guess that calls for some updates there. :slightly_smiling_face:

**Review by crtrott — CHANGES_REQUESTED** [review]
I want to think about this a bit more, in particular I want to think about whether we can make this less overhead. I am actually a bit concerned about the impact of this on subview construction on CPU (even before this PR)

**Comment by masterleinad** [comment]
> I want to think about this a bit more, in particular I want to think about whether we can make this less overhead. I am actually a bit concerned about the impact of this on subview construction on CPU (even before this PR)

What's the order of magnitude of the runtime we expect subview creation to take?

File: core/src/Kokkos_View.hpp [hunk 3]
```diff
@@ -92,26 +92,56 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
+                  label +
+                  "' doesn't match the compile-time dimensions at position " +
+                  std::to_string(i) + ". The given dimension is " +
+                  std::to_string(new_extents[i]) + " but should be " +
+                  std::to_string(View::static_extent(i)) + ".\n";
+              Kokkos::abort(message.c_str());)
+          KOKKOS_IF_ON_DEVICE(
+              Kokkos::abort(
+                  "The specified run-time dimensions for a Kokkos::View has "
+                  "don't match the compile-time dimensions.");)
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:130]: typo "has don't" -> "does not"

File: core/src/Kokkos_View.hpp [hunk 4]
```diff
@@ -92,26 +92,56 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
+    // extents for the static dimensions must match the
+    // compile-time extents.
+    const bool n_args_is_dyn_rank = num_passed_args == dyn_rank;
+    const bool n_args_is_rank     = num_passed_args == rank;
+
+    if (n_args_is_rank) {
+      size_t new_extents[8] = {i0, i1, i2, i3, i4, i5, i6, i7};
+      for (size_t i = dyn_rank; i < rank; ++i)
+        if (new_extents[i] != View::static_extent(i)) {
+          ranks_matching = false;
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:117]:
```suggestion
```

File: core/src/Kokkos_View.hpp [hunk 5]
```diff
@@ -92,26 +92,56 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
+          KOKKOS_IF_ON_HOST(
+              const std::string message =
+                  "The specified run-time dimension for Kokkos::View '" +
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:120]: The specified run-time extent for Kokkos::View '`label`' does not match the compile-time extent along the `i`th dimension.

File: core/unit_test/TestViewCtorDimMatch.hpp [hunk 6]
```diff
@@ -47,384 +47,150 @@
...
+  ASSERT_DEATH(                                                               \
+      EXPR,                                                                   \
+      "Constructor for Kokkos::View 'v' has mismatched number of arguments. " \
+      "The number of arguments = " +                                          \
+          std::to_string(ARGS) +                                              \
+          " neither matches the dynamic rank = " + std::to_string(DYNRANK) +  \
+          " nor the total rank = " + std::to_string(TOTALRANK))
+
+template <int rank, int dynrank, template <int> class RankType,
+          std::size_t... args>
```
Comment (dalg24) [inline core/unit_test/TestViewCtorDimMatch.hpp:61]:
```suggestion
          std::size_t... Is>
```

File: core/unit_test/TestViewCtorDimMatch.hpp [hunk 7]
```diff
@@ -47,384 +47,150 @@
...
-#define PARAM_2 1, 1
-#define PARAM_3 1, 1, 1
-#define PARAM_4 1, 1, 1, 1
-#define PARAM_5 1, 1, 1, 1, 1
-#define PARAM_6 1, 1, 1, 1, 1, 1
-#define PARAM_7 1, 1, 1, 1, 1, 1, 1
+template <int rank, int dynrank, template <int> class RankType>
+void test_matching_arguments_rank() {
+  test_matching_arguments_rank_helper<rank, dynrank, RankType>(
+      std::index_sequence<>());
```
Comment (dalg24) [inline core/unit_test/TestViewCtorDimMatch.hpp:74]:
```suggestion
      std::make_index_sequence<N>());
```

Comment (dalg24) [reply core/unit_test/TestViewCtorDimMatch.hpp:74]: You'd need to use a fold expression to make them all 1
(Not critical change if you dont agree)

File: core/unit_test/TestViewCtorDimMatch.hpp [hunk 8]
```diff
@@ -47,384 +47,150 @@
...
+template <>
+struct DynamicRank<0> {
+  using type = int;
+};
 
 // Skip test execution when KOKKOS_ENABLE_OPENMPTARGET is enabled until
 // Kokkos::abort() aborts properly on that backend
 // Skip test execution when KOKKOS_COMPILER_NVHPC until fixed in GTEST
-#if defined(KOKKOS_ENABLE_OPENMPTARGET) || (KOKKOS_COMPILER_NVHPC)
+#if defined(KOKKOS_ENABLE_OPENMPTARGET) || defined(KOKKOS_COMPILER_NVHPC)
```
Comment (dalg24) [inline core/unit_test/TestViewCtorDimMatch.hpp:106]: I suspect the OpenMPTarget guards are not needed and that we can compile and just skip execution (if needed at all)

File: core/unit_test/TestViewAPI.hpp [hunk 9]
```diff
@@ -1051,7 +1051,7 @@ class TestViewAPI {
 
   static void run_test_contruction_from_layout_2() {
     using dView3_0 = Kokkos::View<T ***, device>;
-    using dView3_1 = Kokkos::View<T * * [N1], device>;
+    using dView3_1 = Kokkos::View<T * * [N2], device>;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/TestViewAPI.hpp:1054]: I added a sentence.

**Comment by masterleinad** [comment]
@crtrott For the benchmark in https://github.com/kokkos/kokkos/issues/5581#issue-1415470255, I see
```
Normal: 1.68759 s.    Unmanaged: 0.622187 s.
```
before this pull request and
```
Normal: 1.70023 s.    Unmanaged: 0.622541 s.
```
after.

**Comment by masterleinad** [comment]
With the change in the other pull request, I am even seeing
```
Normal: 0.642024 s.    Unmanaged: 0.623761 s.
```
(before)
```
Normal: 0.627473 s.    Unmanaged: 0.624141 s.
```
(this pull request).

File: core/src/Kokkos_View.hpp [hunk 10]
```diff
@@ -97,26 +97,53 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
          (i6 != KOKKOS_INVALID_INDEX) + (i7 != KOKKOS_INVALID_INDEX);
 }
 
-KOKKOS_INLINE_FUNCTION
-void runtime_check_rank(const size_t rank, const size_t dyn_rank,
-                        const bool is_void_spec, const size_t i0,
-                        const size_t i1, const size_t i2, const size_t i3,
-                        const size_t i4, const size_t i5, const size_t i6,
-                        const size_t i7, const std::string& label) {
+template <typename View>
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:100]: Add a big fat FIXME that ideally we don't instantiate that guy a whole bunch of times.
If we already required MDSPAN we could just pass `extent`

File: core/src/Kokkos_View.hpp [hunk 11]
```diff
@@ -97,26 +97,53 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
+KOKKOS_INLINE_FUNCTION void runtime_check_rank(
+    const View&, const size_t rank, const size_t dyn_rank,
+    const bool is_void_spec, const size_t i0, const size_t i1, const size_t i2,
+    const size_t i3, const size_t i4, const size_t i5, const size_t i6,
+    const size_t i7, const std::string& label) {
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:105]: I know you did not change that part but how does that even work a `std::string` argument on the device?

File: core/unit_test/TestViewAPI.hpp [hunk 12]
```diff
@@ -1033,25 +1033,25 @@ class TestViewAPI {
     hView3 hv_3("dView3::HostMirror", N0);
     hView4 hv_4("dView4::HostMirror", N0);
 
-    dView0 dv_0_1(nullptr, 0);
+    dView0 dv_0_1(nullptr);
```
Comment (dalg24) [inline core/unit_test/TestViewAPI.hpp:1036]: Why did you change?

Comment (masterleinad) [reply core/unit_test/TestViewAPI.hpp:1036]: Because this is failing with these changes
```
Constructor for Kokkos::View 'UNMANAGED' has mismatched number of arguments. The number of arguments = 1 neither matches the dynamic rank = 0 nor the total rank = 0
2: 
2: Backtrace:
2:  0 KokkosCore_UnitTest_Serial2 0x00000001016d3fd8                                                                                                                                                                                                                                                                                                                                                     Kokkos::Impl::save_stacktrace() +   24
2:  1 KokkosCore_UnitTest_Serial2 0x00000001016cce24                                                                                                                                                                                                                                                                                     Kokkos::Impl::traceback_callstack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) +   36
2:  2 KokkosCore_UnitTest_Serial2 0x00000001016ccee0                                                                                                                                                                                                                                                                                                                                               Kokkos::Impl::host_abort(char const*) +   40
2:  3 KokkosCore_UnitTest_Serial2 0x0000000100fec074                                                                                                                                                                                                                                                                                                                                                              __clang_call_terminate +    0
2:  4 KokkosCore_UnitTest_Serial2 0x00000001010db258 void Kokkos::Impl::runtime_check_rank<Kokkos::View<double, Kokkos::Serial> >(Kokkos::View<double, Kokkos::Serial> const&, unsigned long, unsigned long, bool, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 1292
2:  5 KokkosCore_UnitTest_Serial2 0x00000001010b5db8                                                                                                                                                                                                                                                                                                       Test::TestViewAPI<double, Kokkos::Serial>::run_test_contruction_from_layout() +  736
2:  6 KokkosCore_UnitTest_Serial2 0x00000001010b5094                                                                                                                                                                                                                                                                                                                                            Test::serial_view_api_b_Test::TestBody() +   24
2:  7 KokkosCore_UnitTest_Serial2 0x0000000101691864                                                                                                                                                                                                                                            void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) +  100
2:  8 KokkosCore_UnitTest_Serial2 0x00000001016917b0                                                                                                                                                                                                                                                                                                                                                                testing::Test::Run() +  452
2:  9 KokkosCore_UnitTest_Serial2 0x0000000101692e44                                                                                                                                                                                                                                                                                                                                                            testing::TestInfo::Run() +  480
2: 10 KokkosCore_UnitTest_Serial2 0x0000000101693880                                                                                                                                                                                                                                                                                                                                                           testing::TestSuite::Run() +  356
2: 11 KokkosCore_UnitTest_Serial2 0x00000001016a45b8                                                                                                                                                                                                                                                                                                                                      testing::internal::UnitTestImpl::RunAllTests() + 2144
2: 12 KokkosCore_UnitTest_Serial2 0x00000001016a3bb4                                                                                                                                                                                      bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) +  100
2: 13 KokkosCore_UnitTest_Serial2 0x00000001016a3b1c                                                                                                                                                                                                                                                                                                                                                            testing::UnitTest::Run() +  124
2: 14 KokkosCore_UnitTest_Serial2 0x0000000100feb818                                                                                                                                                                                                                                                                                                                                                                                main +   52
2: 15                        dyld 0x000000010212d08c                                                                                                                                                                                                                                                                                                                                                                               start +  520
1/1 Test #2: KokkosCore_UnitTest_Serial2 ......Subprocess aborted***Exception:   9.32 sec
```

Comment (dalg24) [reply core/unit_test/TestViewAPI.hpp:1036]: OK.  I saw the definition of the dViewX type now.  Makes sense.

File: core/src/Kokkos_View.hpp (hunk 11 above)
Comment (masterleinad) [reply core/src/Kokkos_View.hpp:105]: I guess the compiler ignores `label` on the device since it's a reference and we don't do any operations with it?

**Comment by dalg24** [comment]
Does not build on the GPU backends

**Comment by masterleinad** [comment]
> Does not build on the GPU backends

Should work now.

**Comment by dalg24** [comment]
Retest this please

**Comment by ajpowelsnl** [comment]
> Retest this please

@masterleinad -- would you please re-test?

**Comment by masterleinad** [comment]
> @masterleinad -- would you please re-test?

Damien's comment was for the CI to run again.

**Comment by ajpowelsnl** [comment]
> > @masterleinad -- would you please re-test?
> 
> Damien's comment was for the CI to run again.

OK, my concern was the failing test (perhaps part of the CI issues mentioned recently?)

**Comment by masterleinad** [comment]
> OK, my concern was the failing test (perhaps part of the CI issues mentioned recently?)

No, the CI was last run on this a couple of months ago but your comment retriggered it just now. 🙂

File: core/src/Kokkos_View.hpp [hunk 13]
```diff
@@ -1402,21 +1432,29 @@ class View : public ViewTraits<DataType, Properties...> {
...
+      size_t i2 = arg_layout.dimension[2];
+      size_t i3 = arg_layout.dimension[3];
+      size_t i4 = arg_layout.dimension[4];
+      size_t i5 = arg_layout.dimension[5];
+      size_t i6 = arg_layout.dimension[6];
+      size_t i7 = arg_layout.dimension[7];
+
+      const std::string& alloc_name =
+          Impl::get_property<Impl::LabelTag>(prop_copy);
+      Impl::runtime_check_rank(
```
Comment (fnrizzi) [inline core/src/Kokkos_View.hpp:1453]: as per slack suggestion: maybe we could enable this check in debug build only?

File: core/src/Kokkos_View.hpp [hunk 14]
```diff
@@ -75,25 +75,55 @@ constexpr KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
...
     const size_t num_passed_args =
         count_valid_integers(i0, i1, i2, i3, i4, i5, i6, i7);
+    // We either allow to pass as many extents as the dynamic rank is, or
+    // as many extents as the total rank is. In the latter case, the given
+    // extents for the static dimensions must match the
+    // compile-time extents.
+    const bool n_args_is_dyn_rank = num_passed_args == dyn_rank;
+    const bool n_args_is_rank     = num_passed_args == rank;
+
+    if (n_args_is_rank) {
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:98]: maybe add a `if constexpr (View::rank()!=View::rank_dynamic())` around this?

**Comment by masterleinad** [comment]
`GCC-8.4.0` is timing out, everything else passed.

**Review by fnrizzi — APPROVED** [review]

File: core/src/Kokkos_View.hpp (hunk 14 above)
Comment (crtrott) [reply core/src/Kokkos_View.hpp:99]: I think I still want the if constexpr around the whole thing. At a minimum you are counting the valid integers unnecessarily I think.

**Review by crtrott — CHANGES_REQUESTED** [review]
still want a if constexpr protection.

**Comment by crtrott** [comment]
the if constexpr may also reduce compile time.

File: core/src/Kokkos_View.hpp (hunk 14 above)
Comment (masterleinad) [reply core/src/Kokkos_View.hpp]: I added the check in 90e8cc9c96c0918b627f9bd5e9ffcce6025dd6c9.

**Comment by dalg24** [comment]
Failures (Clang+CUDA and SYCL) are unrelated.  These correspond to #6698
